### PR TITLE
Tidy up the way vendor files are used

### PR DESCRIPTION
**Pull request by jakecleary**
## Intro

There's a few changes here that will effect developer work-flow, but that ensure git isn't tracking too much, and generally tidy up the project a bit to bring it in line with the recommended way to use composer. Happy to discuss this PR if you feel like it needs to be tweaked! 😄 

#### `/tools`
This has been removed. Any scripts found in here are now pulled in as a dev-dependency and can be run using `php vendor/bin/{{script-name-here}}`. I might need your help hooking up the two xml config files (`phpcs.xml` & `phpmd.xml`) which I've temporarily moved to the root directory.

#### `/vendor`
This directory is no longer tracked. Any files inside here are pulled in via composer anyway, and git shouldn't be tracking 3rd party projects. We should modify the `docs/install` file to mention that `composer install` needs to be run when setting up the project. If allowing this to be easily installed on shared web hosts is a concern, we could set up a release script that bundles everything into a zip file and include this as a download on the releases page.

#### `.gitignore`
Git now ignore common OS meta files (for safety), the vendor directory, and IntelliJ meta files which can become a huge pain when they conflict with other people's IDEs.

#### `post-update-cmd`
I've removed the `post-update-cmd` calls that delete a bunch of the vendor project files. Now that these aren't tracked by git, and instead pulled in during the setup phase, I don't see why they should be a concern of yours. Let me know if this isn't the case 😄 

## TODO
- [x] Ensure the vendor scripts can be run using the .xml config files in the root.
- [x] Document the change to the way the tools are used in `/docs/tools`.
- [x] Build a script to bundle and zip each release, including the vendor directory.

### CODE REVIEW

**Comment by thekabal** [comment]
Tools:  "I might need your help hooking up the two xml config files (phpcs.xml & phpmd.xml) which I've temporarily moved to the root directory." . I'd rather we find a solution to the xml config files before removing the tools directory. I'm not opposed to removing the tools directory once we have a good place for the config files and an update for the docs/ references on how to run them.

Vendor: " If allowing this to be easily installed on shared web hosts is a concern, we could set up a release script that bundles everything into a zip file and include this as a download on the releases page.". You've hit on the exact cause of the existing structure. :) Let's complete the release and setup scripts (solving the shared hosting problem) before restructuring based on that problem being solved.

Gitignore: I'll take a look.

Post-update-cmd: See vendor. Once we solve the shared web hosting issue, I have no issue moving this part to the install script. (In fact, I'd prefer it).

In summary, I'd merge pretty much all of this once we get the shared hosting provider bit solved. I like! :)

**Comment by jakecleary** [comment]
So we need to do the following:
1. Write a release script that automates creating the bundle that will be offered as a download on the github/website releases page.
2. Ensure that the `vendor/bin` files are set to load any respective XML config files.

Lets keep this PR open then and I'll look into the above issue in the coming week. Thanks for the quick response time!

**Comment by jakecleary** [comment]
It looks like including the config files is as simple as adding an argument in the command.

**phpcs:** `php vendor/bin/phpcs --standard=phpcs.xml classes/`
- Doesn't seem to raise anything despite the verbose option (`-v`) showing it scanning the files.

**phpmd:** `php vendor/bin/phpmd classes/ text phpmd.xml`
- Outputs the following:
    
```
/Users/jakecleary/Projects/tki-fork/classes/Fighters.php:40	Avoid unused local variables such as '$quantity'.
/Users/jakecleary/Projects/tki-fork/classes/Ship.php:80	Avoid unused local variables such as '$tmp_planet'.
/Users/jakecleary/Projects/tki-fork/classes/Ship.php:84	Avoid unused local variables such as '$planet_id'.
```
Does that output seem correct to you? I'll include this change in the documentation as part of this PR if so. 😄 

Regarding their location, generally speaking the root directory is a common place for putting these kinds of config files. It's very common to see `phpunit.xml` in the root of a project (in fact that's where I put mine).

**Comment by jakecleary** [comment]
Hey @thekabal,

The commits need tidying-up/squashing but the basics are here. Let me know what you think.

**Comment by thekabal** [comment]
The output from phpcs and phpmd is correct. I've worked to get phpcs down to no notices/warnings (when using the rules in the XML). PHPMD's output still shows some errors, I hope to solve those in the next few weeks.

**Comment by thekabal** [comment]
I don't love putting the xml files in the root directory. Let me consider for a little while where they should properly go if the tools directory is removed.

**Comment by thekabal** [comment]
Heh, I think I see why you encouraged git flow.. I'd love to cherry pick several of these changes, but its rather difficult to do so.

For example, the editor exclusions really aren't directly related to the tools clean up, and I'm fine with those, so I'd apply them separately.

Going forward, please try to keep PR's less monolithic ;)

**Comment by jakecleary** [comment]
> The output from phpcs and phpmd is correct. I've worked to get phpcs down to no notices/warnings (when using the rules in the XML). PHPMD's output still shows some errors, I hope to solve those in the next few weeks.

Brilliant work! Was just worried I'd broken them 

> I don't love putting the xml files in the root directory. Let me consider for a little while where they should properly go if the tools directory is removed.

How about we put them in `/config`? I have a proposal regarding the configuration that I'll type up shortly, but this would tidy them away and wouldn't disrupt my idea for how to handle config (DB etc).

> Going forward, please try to keep PR's less monolithic ;)

Yes apologies for that! I realised as I was adding more and more things that the PR scope was growing out of control. I'll spilt up this PR into a few separate ones for you, today. It's no hassle.

**Comment by jakecleary** [comment]
Almost done breaking this PR into smaller chunks :)

**Comment by thekabal** [comment]
"How about we put them in /config? I have a proposal regarding the configuration that I'll type up shortly, but this would tidy them away and wouldn't disrupt my idea for how to handle config (DB etc)."

I think config is a perfectly acceptable location for phpcs & phpmd's xml files.
